### PR TITLE
Make "remove" accept regular expressions

### PR DESCRIPTION
**Pull request by sryze**
Make it possible to remove packages by using a regex as package name (see #8486).

1. Find all installed packages that match the argument(s)
2. Also find an exact name match
3. If a there is package matches exactly, remove only that package (e.g. `vcpkg remove zlib`)
4. If no exact match found, use the regex matches if found any (e.g. `vcpkg remove qt5.*`)

Because this operation may potentially be dangerous / unpredictable (e.g. user unexpectedly removes more packages than expected), I think there should be some sort of confirmation. What do you think?

### CODE REVIEW

**Comment by SuperWig** [comment]
Probably best to do the same thing that `upgrade` does?

**Comment by vicroms** [comment]
I have two suggestions:

* **Require a `--match`, `--regex`, `--pattern` or similar flag to enable regex support.**   
  Preserves the default behavior of `remove` while also allowing the new feature. 

* **Require `--no-dry-run` to actually remove packages when using a regex.**  
  Prevents users from accidentally removing packages, same mechanism as `upgrade`.

**Comment by sryze** [comment]
OK, thanks for the feedback. I'll try to add those flags.

**Comment by sryze** [comment]
I've added 2 new flags: `--pattern` and `--no-dry-run`:

* `--no-dry-run` works like @vicroms suggested
* `--pattern` indicates that the package name(s) are regular expressions

If there are no matching packages found, remove will produce an error:

```
Error: cannot find any packages that match the specified pattern.
```

If it finds a match in the DB but the packages are not installed:

```
$ vcpkg remove zli.* --pattern --no-dry-run
The following packages are not installed, so not removed:
    zlib:x86-windows
Package zlib:x86-windows is not installed
Purging package zlib:x86-windows...
Purging package zlib:x86-windows... done
```

If `--no-dry-run` is omitted and the specified package is installed:

```
$ vcpkg remove zli.* --pattern
The following packages will be removed:
    zlib:x86-windows
If you are sure you want to remove the above packages, run this command with the --no-dry-run option.
```

**Comment by ras0219-msft** [comment]
>If it finds a match in the DB but the packages are not installed:
>```
>The following packages are not installed, so not removed:
>    zlib:x86-windows
>```

I think the regex should be run against the list of _installed_ libraries, not against the list of available libraries. Consider the case that a recipe is removed (say, some sublibrary of boost) -- currently, this would make it inaccessible via the pattern remove mechanism of `boost.*`.

In the example above, I believe the pattern form of the command should print that no installed libraries match the pattern `zli.*`.

Otherwise, I approve of this design 👍.

File: toolsrc/src/vcpkg/remove.cpp [hunk 1]
```diff
@@ -188,14 +189,18 @@ namespace vcpkg::Remove
...
     static constexpr StringLiteral OPTION_OUTDATED = "--outdated";
+    static constexpr StringLiteral OPTION_PATTERN = "--pattern";
 
-    static constexpr std::array<CommandSwitch, 5> SWITCHES = {{
+    static constexpr std::array<CommandSwitch, 7> SWITCHES = {{
         {OPTION_PURGE, "Remove the cached copy of the package (default)"},
         {OPTION_NO_PURGE, "Do not remove the cached copy of the package"},
         {OPTION_RECURSE, "Allow removal of packages not explicitly specified on the command line"},
         {OPTION_DRY_RUN, "Print the packages to be removed, but do not remove them"},
+        {OPTION_NO_DRY_RUN, "Actually remove"},
```
Comment (ras0219-msft) [inline toolsrc/src/vcpkg/remove.cpp]: Nit: it is unfortunate that this option has such similar semantics to `--recurse`. I feel that it would be a nice drive-by improvement to make this option imply `--recurse`.

File: toolsrc/src/vcpkg/remove.cpp [hunk 2]
```diff
@@ -188,14 +189,18 @@ namespace vcpkg::Remove
...
         {OPTION_OUTDATED, "Select all packages with versions that do not match the portfiles"},
+        {OPTION_PATTERN, "Select all packages that match a regular expression pattern"},
```
Comment (ras0219-msft) [inline toolsrc/src/vcpkg/remove.cpp]: We should document the regex flavor in use (ECMAScript, if I recall correctly).

File: toolsrc/src/vcpkg/remove.cpp [hunk 3]
```diff
@@ -247,10 +255,47 @@ namespace vcpkg::Remove
...
+                {
+                    Optional<PackageSpec> exact_spec;
+                    std::vector<PackageSpec> regex_specs;
+                    std::regex arg_regex{arg};
+                    for (const auto& status : status_db) 
+                    {
+                        const auto& spec_name = Strings::ascii_to_lowercase(std::string(status->package.spec.name()));
+                        if (spec_name == arg)
+                        {
+                            // Found an exact match, no need to search further.
```
Comment (ras0219-msft) [inline toolsrc/src/vcpkg/remove.cpp]: In pattern mode, this should always search the entire list. I believe we currently disallow all special regex characters in port names, so this implementation is currently correct, but the implementation would be cleaner without this shortcut.

**Comment by dan-shaw** [comment]
@sryze Is this PR still being worked on? Can you address the review comments? This is a really important PR!

**Comment by JackBoosY** [comment]
@dan-shaw Can we create a new PR based on this PR to deal with this feature?

**Comment by dan-shaw** [comment]
Closing PR due to inactivity. We think this feature is a good idea, so we are still accepting PRs for this feature.
